### ISSUE TITLE
Cannot stop simulated process target

### ISSUE THREAD

**Issue by hth313** [microsoft/vscode-cpptools#334]
I am using a simulation environment where threads and processes are internal to the debugger, hence thread and thread group identities are *not* valid outside the debugger process.

After launch I cannot stop the debugee. The pause button has no effect.

Looking at MICore, it seems that it wants to send `SIGINT` to the debugee process, not the debugger process.  But it is not possible to send signals to the simulated debugee process from the outside in the setup I have. The debugger can handle `^C`, `-exec-interrupt` and the console command `interrupt`.

Is there a way to convince vscode-cpptools to use `-exec-interrupt`?

I am using a launch setup on Linux (I also have access to Mac OS X).

**Comment by pieandcakes** [comment]
@hth313 If I remember correctly, this was changed because of an issue with attaching to a process versus launch and how the debugger handled it. This might be an issue that needs to be filed at the MIEngine [issues](http://github.com/microsoft/miengine/issues) page or if you want, you can download that source code and make a change to support this specific scenario. 

I'll mark this as a feature request and we can see if there might be something we can do.

**Comment by hth313** [comment]
@pieandcakes Using  `-exec-interrupt` is the correct way to stop a debugee with MI. Why would the command even exist otherwise?

This allows the debugger to take whatever specific means there are to interrupt execution. If a debugger has problems with `-exec-interrupt`, I would say there is a bug at the debugger end.

I am really stuck on this and will put this aside for a while. If you have any description on how to set up MIEngine to build it and how to replace it inside vscode/cpp-tools, it could be useful. 

Even if I poke around and enable the code again (which probably is simple in a proper build setup), I wonder if you will to accept it, as this change was done intentional. Maybe there is a third way, we'll see.

**Comment by pieandcakes** [comment]
@hth313 Using `SIGINT` was a workaround to an issue in the debugger. I believe the person working on the issue logged a bug against gdb but I wasn't able to find it through a quick search. 

The instructions to download MIEngine code and build is [here](https://github.com/Microsoft/MIEngine#get-started) on their GitHub page. The build definition you want to choose is `Desktop.Release`. You will need Visual Studio to build and after the solution is built, copy the following files into `~/.vscode/extensions/ms-vscode.cpptools-0.9.2/debugAdapters/bin` folder:
- Microsoft.MICore.dll
- Microsoft.MIDebugEngine.dll
- Microsoft.MICore.XmlSerializers.dll

**Comment by hth313** [comment]
@pieandcakes Thanks! Unfortunately it says I need to use Visual Studio on Windows, which means I will not be able to do it.

I tried to find the bug report as well, but could not locate it.

**Comment by hth313** [comment]
@pieandcakes  I noticed that it is only gdb that does this, so I switched ```MIMode``` to ```lldb```, but still I could not stop it. Then I updated to VS Code 1.7.2 and finally the pause button had some effect. 

It seems that I now can interrupt execution as indicated on the console, but not much else is happening. I cannot even issue any manual commands to the debugger using ```-exec info registers```, it just responds with: 

```error while processing request 'evaluate' (exception: Specified argument was out of the range of valid values.\nParameter name: handle)```

Continuing (or stepping) execution does not seem to work either from the interrupted state, so I suspect it is still confused about the execution status.

**Comment by pieandcakes** [comment]
@hth313 There was an issue [#15165](https://github.com/Microsoft/vscode/issues/15165) in 1.7.1 that caused the pause button not to work. If you are using lldb, `-exec info registers` is not a valid command. You will need to find the correct lldb command which for the above case should be `-exec register read`. 

Would you be able to provide sample code for the break and then step not working? with my test app I can't seem to duplicate it.

**Comment by hth313** [comment]
@pieandcakes Once stopped, the ```-exec``` commands does not seem to be delivered to the debugger at all. (The commands I try to send are known to the debugger.)

Here is what happens:

```
-exec info reg
error while processing request 'evaluate' (exception: Specified argument was out of the range of valid values.\nParameter name: handle)
C evaluate: {"expression":"-exec info reg","context":"repl"}
 R: {"success":false,"message":"error while processing request 'evaluate' (exception: Specified argument was out of the range of valid values.\nParameter name: handle)","request_seq":9,"command":"evaluate","body":{"error":{"id":1104,"format":"error while processing request '{_request}' (exception: {_exception})","variables":{"_request":"evaluate","_exception":"Specified argument was out of the range of valid values.\nParameter name: handle"}}},"running":false,"refs":null,"seq":0,"type":"response"}
```

I tried to ```find-grep``` around in MIEngine and Visual Studio Code, but I could not locate the code where it happens, so I assume it is someplace else.

Is there some description of the console command line interface somewhere?

**Comment by pieandcakes** [comment]
@hth313 that looks like a bug. We are expecting the `evaluate: ` line to have frameId which should be the stack frame id to evaluate the expression against, but for some reason in this instance it has no frameId which is causing an exception to be thrown. 

Can you verify that you have a callstack and if so, can you send me the logging messages after the break that show it retrieving the callstack?

**Comment by hth313** [comment]
@pieandcakes It is an assembler project, so it does not have stack frames the normal way. It can report a frame 0 with the current address when asked to, see below.

Here is the complete output I have in VS Code, starting it and then pressing pause:
```
E initialized: {"type":"initialized"}
E output: {"category":"telemetry","output":"VS/Diagnostics/Debugger/Launch","data":{"VS.Diagnostics.Debugger.ImplementationName":"Microsoft.MIDebugEngine","VS.Diagnostics.Debugger.EngineVersion":"14.0.30917.1","VS.Diagnostics.Debugger.HostVersion":"1.0.20917.1","VS.Diagnostics.Debugger.AdapterId":"cppdbg","VS.Diagnostics.Debugger.Launch.Duration":397,"VS.Diagnostics.Debugger.Launch.IsCoreDump":false},"type":"output"}
 R: {"success":true,"message":null,"request_seq":2,"command":"launch","body":null,"running":false,"refs":null,"seq":0,"type":"response"}
C setFunctionBreakpoints: {"breakpoints":[]}
 R: {"success":true,"message":null,"request_seq":3,"command":"setFunctionBreakpoints","body":{"breakpoints":[]},"running":false,"refs":null,"seq":0,"type":"response"}
C configurationDone: null
 R: {"success":true,"message":null,"request_seq":4,"command":"configurationDone","body":null,"running":false,"refs":null,"seq":0,"type":"response"}
E output: {"category":"stdout","output":"copying output to \"dbnut.txt\"\nlistening for user interface connections on port 8080\nWarning: Debuggee TargetArchitecture not detected, assuming x86_64.\n","data":null,"type":"output"}
copying output to "dbnut.txt"
listening for user interface connections on port 8080
E output: {"category":"telemetry","output":"VS/Diagnostics/Debugger/Pause","data":{"VS.Diagnostics.Debugger.ImplementationName":"Microsoft.MIDebugEngine","VS.Diagnostics.Debugger.EngineVersion":"14.0.30917.1","VS.Diagnostics.Debugger.HostVersion":"1.0.20917.1","VS.Diagnostics.Debugger.AdapterId":"cppdbg"},"type":"output"}
E output: {"category":"console","output":"Execute debugger commands using \"-exec <command>\", for example \"-exec info registers\" will list registers in use (when GDB is the debugger)\n","data":null,"type":"output"}
Execute debugger commands using "-exec <command>", for example "-exec info registers" will list registers in use (when GDB is the debugger)
E stopped: {"threadId":42,"reason":"pause","source":null,"line":0,"column":0,"text":null,"allThreadsStopped":true,"type":"stopped"}
C threads: null
 R: {"success":true,"message":null,"request_seq":6,"command":"threads","body":{"threads":[]},"running":false,"refs":null,"seq":0,"type":"response"}
```

Here is the MI log as seen by the debugger:
```
~"listening for user interface connections on port 8080\n"
(dbnut) 1001-gdb-set solib-search-path "/home/hth/projects/module/ladybug/src:"
1001^done
(gdb) 
(dbnut) 1002-environment-cd /home/hth/projects/module/ladybug/src
1002^done
(gdb) 
(dbnut) 1003-file-exec-and-symbols /home/hth/projects/module/ladybug/src/ladybug
1003^done
(gdb) 
(dbnut) 1004-interpreter-exec console "platform status"
~"target architecture is set automatically (currently nut)\n"
1004^done
(gdb) 
(dbnut) 1005-exec-arguments --load-module /home/hth/projects/modules41/41cv.mod
1005^done
(gdb) 
(dbnut) 1006-break-insert main
1006^error,msg="unknown label: main"
(gdb) 
(dbnut) 1007-exec-interrupt
1007^done
(gdb) 
1007*stopped,reason="signal-received",signal-name="SIGINT",signal-meaning="interrupt",frame={level="0",addr="0x0000"},thread-id="1",stopped-threads="1",core="0"
(dbnut) 1008-thread-info
1008^done,threads=[{id="1",target-id="process 42",state="stopped",core="0",frame={level="0",addr="0x0000"}}],current-thread-id="1"
(gdb) 
(dbnut) 1009-stack-list-frames 0 1000 --thread 1
1009^done,stack=[frame={level="0",addr="0x0000"}]
(gdb) 
```

**Comment by pieandcakes** [comment]
@hth313 Can you shoot me an email at piel at microsoft dot com? I can give you a proposed fix to see if it might resolve your issue.

**Comment by pieandcakes** [comment]
Resolved this with @hth313 over email. The fixes are in place in `0.10.2`
